app/lib/reset_tenant_id.rb

```diff
@@ -0,0 +1,144 @@
+module ResetTenantId
+  extend ActiveSupport::Concern
+
+  class Configurator
+    include Singleton
+    attr_reader :definitions
+
+    def initialize
+      @definitions = {}
+    end
+
+    def define(klass,  &block)
+      definition = definitions[klass] ||= Definition.new(klass)
+      definition.instance_exec(&block)
+    end
+  end
+
+  class Definition
+    attr_reader :queries
+    delegate :each_value, to: :queries
+
+    def initialize(klass)
+      @klass = klass
+      @queries = {}
+    end
+
+    def updates(name, options = {}, &block)
+      @queries[name.to_sym] = Query.new(name, options, &block)
+    end
+
+    def [](name)
+      @queries[name.to_sym]
+    end
+  end
+
+  class Query
+    attr_reader :collection, :name, :with
+
+    def initialize(name, options={}, &block)
+      @name = name.to_sym
+      @collection = options[:collection] || @name
+      @with = options[:with] || block
+    end
+
+    # Returns an object responding to #call(record)
+    def update
+      case @with
+      when Symbol
+        Proc.new do |record|
+          record.update_column :tenant_id, record.public_send(@with)
+        end
+      else
+        @with
+      end
+    end
+  end
+
+  class Runner
+    attr_reader :klass, :definitions
+
+    def initialize(klass, definitions)
+      @klass = klass
+      @definitions = definitions
+    end
+
+    def execute(name)
+      query = find_query(name)
+      # probably adding logging here
+      return unless query
+      execute_query(query)
+    end
+
+    def execute_all
+      definitions[klass].each_value do |query|
+        execute_query(query)
+      end
+    end
+
+    def execute_query(query)
+      # This can be hooked to be async :)
+      collection = query.collection
+      collection = case collection
+                   when Proc
+                     klass.instance_exec(&collection)
+                   when Symbol
+                     klass.public_send(collection)
+                   when Array
+                     collection.extends(FindEach)
+                   else
+                     collection
+                   end
+
+      collection.find_each(&query.update)
+    end
+
+    def find_query(name)
+      definitions[klass][name.to_sym]
+    end
+
+    module FindEach
+      def find_each(&block)
+        defined?(super) ? super(&block) : each(&block)
+      end
+    end
+  end
+
+  def reset_tenant_id!(name)
+    runner = ResetTenantId::Runner.new(self.class, ResetTenantId::Configurator.instance.definitions)
+    query = runner.find_query(name)
+    query.update.call(self) unless query
+  end
+
+  module ClassMethods
+
+    # define_reset_tenant_id do
+    #   updates(:buyers, with: :provider_account_id)
```

Comment:
Not sure if `with` is self-explanatory. It could also be `using` or `source`.

Reply:
well it should be english-ish readable.

`updates buyers source provider_account_id` is not really english
`updates buyers using provider_account_id` is OK but how is it different from
`updates buyers with provider_account_id`

Reply:
It's a good point. My concern because it's not clear what the `provider_account_id` is for. I mean we know this is where we get the value from to set `tenant_id`, but does `with` say that? Maybe there's nothing better. IDK.